File: csharp/ToolGood.Algorithm/Internals/Functions/CsharpWeb/Function_BASE64TOTEXT.cs
using System;
using System.Collections.Generic;
using System.Text;
using ToolGood.Algorithm.Enums;

namespace ToolGood.Algorithm.Internals.Functions.CsharpWeb
{
	/// <summary>
	/// Base64ToText：将标准 Base64 字符串按 UTF-8 解码为文本。
	/// 输入不符合 Base64 格式时返回参数错误。
	/// </summary>
	internal sealed class Function_BASE64TOTEXT : Function_1
	{
		public Function_BASE64TOTEXT(FunctionBase[] funcs) : base(funcs)
		{
			if (funcs.Length != 1) {
				throw new ArgumentException($"Function '{Name}' requires exactly 1 parameter.");
			}
		}

		public override string Name => "Base64ToText";

		public override Operand Evaluate(AlgorithmEngine engine, Func<AlgorithmEngine, string, Operand> tempParameter)
		{
			var args1 = GetText_1(engine, tempParameter);
			if(args1.IsErrorOrNone) { return args1; }
			try {
				var t = Encoding.UTF8.GetString(Convert.FromBase64String(args1.TextValue));
				return Operand.Create(t);
			} catch {
				return ParameterError(1);
			}
		}
		public override OperandType GetResultType()
		{
			return OperandType.TEXT;
		}

		internal override void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, string op = null, string val = null)
		{
			func1.GetParameterTypes(noneEngine, result, OperandType.TEXT);
		}
	}

}

File: java/src/main/java/toolgood/algorithm/internals/functions/csharpweb/Function_BASE64TOTEXT.java
package toolgood.algorithm.internals.functions.csharpweb;

import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.List;
import java.util.function.BiFunction;

import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_1;
import toolgood.algorithm.internals.functions.NoneEngine;

public final class Function_BASE64TOTEXT extends Function_1 {
    public Function_BASE64TOTEXT(FunctionBase func1) {
        super(func1);
    }

    @Override
    public String Name() {
        return "Base64ToText";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        Operand args1 = GetText_1(engine, tempParameter);
        if (args1.IsErrorOrNone()) {
            return args1;
        }
        try {
            byte[] bytes = Base64.getDecoder().decode(args1.TextValue());
            String t = new String(bytes, StandardCharsets.UTF_8);
            return Operand.Create(t);
        } catch (Exception e) {
            return ParameterError(1);
        }
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.TEXT;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType,
            String op, String val) {
        func1.GetParameterTypes(noneEngine, result, OperandType.TEXT);
    }
}
